addons/sale/models/sale.py

```diff
@@ -259,10 +259,15 @@ def onchange_partner_id_warning(self):
     @api.model
     def create(self, vals):
         if vals.get('name', _('New')) == _('New'):
-            if 'company_id' in vals:
-                vals['name'] = self.env['ir.sequence'].with_context(force_company=vals['company_id']).next_by_code('sale.order') or _('New')
-            else:
-                vals['name'] = self.env['ir.sequence'].next_by_code('sale.order') or _('New')
+            ctx = self.env.context.copy()
+            if vals.get('company_id'):
+                ctx['force_company'] = vals['company_id']
+            if vals.get('date_order'):
+                ctx['ir_sequence_date'] = self._fields['date_order'].convert_to_cache(vals['date_order'], False, validate=False)[:10]
```

Comment:
This is ugly: do we need to pass a string or a date object ?

Reply:
Explanation in later comment